--- autoload.php
+++ autoload.php
@@ -23,6 +23,7 @@ if ( !defined( 'EZCBASE_ENABLED' ) )
 {
     $appName = defined( 'EZP_APP_FOLDER_NAME' ) ? EZP_APP_FOLDER_NAME : 'ezpublish';
     $appFolder = __DIR__ . "/../$appName";
+    $legacyVendorDir = __DIR__ . "/vendor";

     // Bundled
     if ( defined( 'EZP_USE_BUNDLED_COMPONENTS' ) ? EZP_USE_BUNDLED_COMPONENTS === true : file_exists( __DIR__ . "/lib/ezc" ) )
@@ -43,6 +44,12 @@ if ( !defined( 'EZCBASE_ENABLED' ) )
         require_once "{$appFolder}/autoload.php";
         $baseEnabled = false;
     }
+    // Composer if in eZ Publish legacy context
+    else if ( file_exists( "{$legacyVendorDir}/autoload.php" ) )
+    {
+        require_once "{$legacyVendorDir}/autoload.php";
+        $baseEnabled = false;
+    }
     // PEAR install
     else
     {

--- bin/php/ezpgenerateautoloads.php
+++ bin/php/ezpgenerateautoloads.php
@@ -18,6 +18,7 @@
 //{
 $appName = defined( 'EZP_APP_FOLDER_NAME' ) ? EZP_APP_FOLDER_NAME : 'ezpublish';
 $appFolder = getcwd() . "/../$appName";
+$legacyVendorDir = getcwd() . "/vendor";
 
 $baseEnabled = true;
 // Bundled
@@ -37,6 +38,12 @@
     require_once "{$appFolder}/autoload.php";
     $baseEnabled = false;
 }
+// Composer if in eZ Publish legacy context
+else if ( file_exists( $legacyVendorDir . "/autoload.php" ) )
+{
+    require_once $legacyVendorDir . "/autoload.php";
+    $baseEnabled = false;
+}
 // PEAR
 else
 {
